bitcoin/src/taproot/mod.rs

```diff
@@ -576,52 +574,72 @@ impl Default for TaprootBuilder {
     fn default() -> Self { Self::new() }
 }
 
+/// Error if attempting an operation that requires the `TaprootBuilder` to be finalizable.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct IncompleteBuilderError {
+    /// The unmodified incomplete builder.
+    pub builder: TaprootBuilder,
+}
```

Comment:
Yes, this makes a lot of sense.

Reply:
I did some work yesterday where I put all the `hex` error changes on a branch and depended on it in `rust-bitcoin`. Hiding the internals broke `rust-bitcoin`, led me to think that all our struct error types (that exist solely to hide internals) should have private fields and getters for the fields. Way more forwards compatible then.

Reply:
This one doesn't exist solely to hide internals though. And I really doubt we can come up with a better layout in this case.

You're probably right in those other cases.

Reply:
>  can't destructure during pattern matching

They can, they just have to use `..` in every time and that allows us to add new fields.

Reply:
I must have been confused 5 days ago, this error definitely is a hider error, otherwise the `IntoTapTreeError::IncompleteBuilder` variant would just have the `TaprootBuilder` in it directly. And for that matter this PR is incomplete because it leaves `IntoTapTreeError::HiddenParts(TaprootBuilder)` as such, which is inconsistent. Phew this error stuff is hard to get right, I'm finding the challenge more and more interesting.

Reply:
Not really. It's actually good to wrap the thing so that it has a proper error message etc but allows getting the original builder. Actually now that I think about it I don't think there will ever be additional information in the error because the builder already contains everything. I'd still keep `non_exhaustive` as a lint though.

> I'm finding the challenge more and more interesting.

Cool!